Update table.multi.r

markriseley opened the pull request:
Fix false positive matches with grep in multi.split eg "Apple" also matches "Crab Apple".

Minimum reproducible example of previous error (should only be 1 match on "Apple":

      fruits = c("Apple/Pear", "Crab Apple/Peach")
      library(testthat)
      test_that({
          t1 <- as.data.frame(questionr::multi.split(fruits))
          expect_equal(sum(t1[["fruits.Apple"]]),1)
      })

juba commented:
Thanks for reporting !

`%in%` doesn't seem to work here (I get , I think we still need regular expressions. I'll try to push a fix soon.

markriseley commented:
Yup I see what you mean, I'm sure there's a regex for it.

On 13 July 2015 at 16:03, Julien Barnier <notifications@github.com> wrote:

> Thanks for reporting !
>
> %in% doesn't seem to work here (I get , I think we still need regular
> expressions. I'll try to push a fix soon.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/juba/questionr/pull/58#issuecomment-120960521>.
>

briatte commented:
Code below should work. Why did you need `fixed =TRUE` in the first place?

```r
multi.split <- function (var, split.char="/", mnames = NULL) {
  vname <- deparse(substitute(var))
  lev <- levels(factor(var))
  lev <- unique(unlist(strsplit(lev, split.char)))
  if (is.null(mnames)) 
    mnames <- gsub(" ", "_", paste(vname, lev, sep = "."))
  else mnames <- paste(vname, mnames, sep = ".")
  result <- matrix(data = 0, nrow = length(var), ncol = length(lev))
  char.var <- as.character(var)
  for (i in 1:length(lev)) {
    result[grep(paste0("^", lev[i], split.char), char.var), i] <- 1
    result[grep(paste0(split.char, lev[i], split.char), char.var), i] <- 1
    result[grep(paste0(split.char, lev[i], "$"), char.var), i] <- 1
  }
  result <- data.frame(result)
  colnames(result) <- mnames
  result
}
```

```r
fruits = c("Apple/Pear", "Crab Apple/Peach")
multi.split(fruits)
```
```
  fruits.Apple fruits.Pear fruits.Crab_Apple fruits.Peach
1            1           1                 0            0
2            0           0                 1            1
```

markriseley commented:
That was because I didn't want to have to escape "split.char" eg for pipes
`|`. I don't think that's straightforward to handle with the revised code
so don't worry about it.

I will just pre-process the split.char input (was for a shiny app hence not
wanting to ask non-programmers to do regex work).

Thanks for fixing!


On 14 July 2015 at 17:23, François <notifications@github.com> wrote:

> Code below should work. Why did you need fixed =TRUE in the first place?
>
> multi.split <- function (var, split.char="/", mnames = NULL) {
>   vname <- deparse(substitute(var))
>   lev <- levels(factor(var))
>   lev <- unique(unlist(strsplit(lev, split.char)))
>   if (is.null(mnames))
>     mnames <- gsub(" ", "_", paste(vname, lev, sep = "."))
>   else mnames <- paste(vname, mnames, sep = ".")
>   result <- matrix(data = 0, nrow = length(var), ncol = length(lev))
>   char.var <- as.character(var)
>   for (i in 1:length(lev)) {
>     result[grep(paste0("^", lev[i], split.char), char.var), i] <- 1
>     result[grep(paste0(split.char, lev[i], split.char), char.var), i] <- 1
>     result[grep(paste0(split.char, lev[i], "$"), char.var), i] <- 1
>   }
>   result <- data.frame(result)
>   colnames(result) <- mnames
>   result
> }
>
> fruits = c("Apple/Pear", "Crab Apple/Peach")
> multi.split(fruits)
>
>   fruits.Apple fruits.Pear fruits.Crab_Apple fruits.Peach
> 1            1           1                 0            0
> 2            0           0                 1            1
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/juba/questionr/pull/58#issuecomment-121296196>.
>

markriseley commented:
FYI you need one more grep pattern to cover the case of an item in a list
with no split separators eg

fruits = c("Apple/Pear", "Crab Apple/Peach", "*Orange*")

Add this to your results for loop and I think you have it covered:

result[grep(paste0("^",lev[i],"$"), char.var), i] <- 1

On 16 July 2015 at 13:07, Mark Riseley <markriseley@googlemail.com> wrote:

> That was because I didn't want to have to escape "split.char" eg for pipes
> `|`. I don't think that's straightforward to handle with the revised code
> so don't worry about it.
>
> I will just pre-process the split.char input (was for a shiny app hence
> not wanting to ask non-programmers to do regex work).
>
> Thanks for fixing!
>
>
> On 14 July 2015 at 17:23, François <notifications@github.com> wrote:
>
>> Code below should work. Why did you need fixed =TRUE in the first place?
>>
>> multi.split <- function (var, split.char="/", mnames = NULL) {
>>   vname <- deparse(substitute(var))
>>   lev <- levels(factor(var))
>>   lev <- unique(unlist(strsplit(lev, split.char)))
>>   if (is.null(mnames))
>>     mnames <- gsub(" ", "_", paste(vname, lev, sep = "."))
>>   else mnames <- paste(vname, mnames, sep = ".")
>>   result <- matrix(data = 0, nrow = length(var), ncol = length(lev))
>>   char.var <- as.character(var)
>>   for (i in 1:length(lev)) {
>>     result[grep(paste0("^", lev[i], split.char), char.var), i] <- 1
>>     result[grep(paste0(split.char, lev[i], split.char), char.var), i] <- 1
>>     result[grep(paste0(split.char, lev[i], "$"), char.var), i] <- 1
>>   }
>>   result <- data.frame(result)
>>   colnames(result) <- mnames
>>   result
>> }
>>
>> fruits = c("Apple/Pear", "Crab Apple/Peach")
>> multi.split(fruits)
>>
>>   fruits.Apple fruits.Pear fruits.Crab_Apple fruits.Peach
>> 1            1           1                 0            0
>> 2            0           0                 1            1
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/juba/questionr/pull/58#issuecomment-121296196>.
>>
>
>

juba commented:
Sorry for the late reply, as I'm currently on vacation.

Does the problem still exist in `master` ? I thought that this commit might have solved it : 

https://github.com/larmarange/questionr/commit/4eda36cd56d4d59886d6bf598a30e1db930425a4

markriseley commented:
Ah yes, it's solved there.

I was responding to Francois's suggested code which wasn't quite there.

Good stuff, now time for me to get serious learning regex...

On 23 July 2015 at 14:28, Julien Barnier <notifications@github.com> wrote:

> Sorry for the late reply, as I'm currently on vacation.
>
> Does the problem still exist in master ? I thought that this commit might
> have solved it :
>
> larmarange@4eda36c
> <https://github.com/larmarange/questionr/commit/4eda36cd56d4d59886d6bf598a30e1db930425a4>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/juba/questionr/pull/58#issuecomment-124098656>.
>
